### PR TITLE
NAS-111806 / 21.08 / Add tests for directory services user/group cache

### PR DESCRIPTION
**Pull request by anodos325**
During AD join process we generate a list of IDs known to us from
Samba's gencache.tdb, and then iterate the list, converting their
respective pwd and grp entries into our typical user / group query
result structures. These get inserted into key-value stores in
/root/tdb/persistent on the local filesystem. The local cache
gets refilled once per day. Directory service users / groups
are included in query results when the filter-option
`{"extra": {"search_dscache": True}}` is specifid. LDAP server
is not queried directly if it can be avoided so that unnecessary
load is not generated (especially in large or poorly designed
environments). This was historically a major issue in legacy
versions of FreeNAS (11.1 and earlier). This does create an
issue for cases where recently added users may not be present
in the cache. It is common for API users to query a single
user during user / group validation with a single filter
`[["username", "=", <name>]]`. In this case the directory services
name cache will look up the user directly through nss (pwd or grp module).
If user is found, cache entry will be inserted into the cache
backend so that it will appear in future query results.

This series of tests validates the following:
- Cache fill job completes successfully
- Cache refresh job completes successfully
- Direct lookup of user / group by id or name works correctly
- Cache entries are inserted after direct lookup

### CODE REVIEW

**Review by ericbsd — APPROVED** [review]
LGTM

File: tests/api2/test_040_ad_user_group_cache.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    results = GET('/group', payload={
+        'query-filters': [['name', '=', ad_group]],
+        'query-options': {'extra': {"search_dscache": True}},
+    })
+    assert results.status_code == 200, results.text
+    assert len(results.json()) > 0, results.text
+    if len(results.json()) == 0:
+        return
+
+    ad_domain_users_id = results.json()[0]['gid']
```
Comment (rick-mesta) [inline tests/api2/test_040_ad_user_group_cache.py:240]: `ad_domian_users_id` is set, but the `assert()` below is looking at `ad_group` ... same thing ?

File: tests/api2/test_040_ad_user_group_cache.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    results = GET('/user', payload={
+        'query-filters': [['local', '=', False]],
+        'query-options': {'extra': {"search_dscache": True}},
+    })
+    assert results.status_code == 200, results.text
+    assert len(results.json()) > 0, results.text
+
+
+@pytest.mark.dependency(name="AD_GROUPS_CACHED")
+def test_08_check_for_ad_users(request):
```
[the remark this answers is not in the document]
Comment (anodos325) [reply tests/api2/test_040_ad_user_group_cache.py]: Looks like you had an old version. Perhaps github hiccup. Fixed.

**Review by rick-mesta — APPROVED** [review]
Thx for the detailed explanation and _*most excellent*_  job on these functional tests 👍🏼

**Pull request merged**